librbd: don't set objectmap invalid when load objectmap timed out

Songweibin opened the pull request:
return negative error code and don't set objectmap invalid when load objectmap timed out.

Signed-off-by: songweibin <song.weibin@zte.com.cn>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

dillaman commented on src/librbd/object_map/RefreshRequest.cc:
```diff
@@ -133,6 +133,10 @@ Context *RefreshRequest<I>::handle_load(int *ret_val) {
     return nullptr;
   } else if (*ret_val < 0) {
     lderr(cct) << "failed to load object map: " << oid << dendl;
+    if (*ret_val == -ETIMEDOUT) {
```
You will need to invalidate the object-map at this point (if it's for the HEAD revision)

Songweibin replied:
backgroud info: in iSCSI scenario, set `rados_osd_op_timeout = 10`
1. create a test image `lun1`
2. client A opens and writes to image `lun1`
3. client B opens image `lun1`
4. use `tc` to add a delay to a storage node C
5. client A write timed out and close image `lun1` 
6. client B try to acquire exclusive lock and failed (the object-map of `lun1` happens to be at node C)
7. client B successfully acquired exclusive lock, but the object-map is set to invalid

we think in this scenario, object-map should not set to invalid and return error to client B.
not sure if this is the right way to reslove it.

@dillaman  what do you think? thanks.
```
2020-03-19 11:13:15.292862 7f60177fe700  5 librbd::ManagedLock: 0x7f600800ac30 handle_acquire_lock: successfully acquired exclusive lock
2020-03-19 11:13:15.292868 7f60177fe700 10 librbd::ExclusiveLock: 0x7f600800ac30 post_acquire_lock_handler: r=0
2020-03-19 11:13:15.292876 7f60177fe700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 send_open_object_map:
2020-03-19 11:13:15.292885 7f60177fe700 20 librbd::object_map::RefreshRequest: 0x7f6010004570 send: object_count=524288
2020-03-19 11:13:15.292891 7f60177fe700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_lock: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:15.292894 7f60177fe700 10 librbd::object_map::LockRequest: 0x7f6010001ac0 send_lock: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:16.294808 7f6017fff700 10 librbd::object_map::LockRequest: 0x7f6010001ac0 handle_lock: r=0
2020-03-19 11:13:16.294821 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_lock
2020-03-19 11:13:16.294827 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_load: oid=rbd_object_map.1afb6b8b4567
2020-03-19 11:13:26.294961 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_load: r=-110
2020-03-19 11:13:26.294979 7f6017fff700 -1 librbd::object_map::RefreshRequest: failed to load object map: rbd_object_map.1afb6b8b4567
2020-03-19 11:13:26.294987 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 send_invalidate
2020-03-19 11:13:26.327480 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 invalidating object map in-memory
2020-03-19 11:13:26.327487 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 invalidating object map on-disk
2020-03-19 11:13:36.327620 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 should_complete: r=-110
2020-03-19 11:13:36.327635 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_invalidate: r=0
2020-03-19 11:13:36.327647 7f6017fff700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 handle_open_object_map: r=0
```

dillaman replied:
If you cannot open the object map, that means you won't update the object-map, which means that it needs to be invalidated. If your OSDs are timing out after 10 seconds, perhaps object-map is the least of your concerns.

Why, in your example, does the load time-out when client B opens the object-map?

Songweibin replied:
@dillaman  If time-out occurs when open the objectmap and return error code to APPs, we have reason to suspect that there may be a problem with the network, but if time-out and returns to normal, we donot necessarily know that it is caused by the network.

I think we should try to avoid setting the object map to invalid. There are two main considerations: first, many RBD operations are based on the object map, which will be time-consuming and inefficient without it; second, it may take a long time to rebuild a large image (eg. 10TB) which will affect the business.

In addition, if invalidating failed, the opened image will run in the no object map state, but at the same time it can continue writing. Actually, object map on the server is still valid, which will cause the mismatch between the object map and the actual image object and may cause image damage.

> Why, in your example, does the load time-out when client B opens the object-map?

Cause the object map of `lun1` happens to be at node C which its network is in sub-health state (we have this test case).

Thank you.

dillaman replied:
We shouldn't return an error to the app on IO -- and I don't see that in your appended log section:

> 2020-03-19 11:13:36.327647 7f6017fff700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 handle_open_object_map: r=0

dillaman replied:
> In addition, if invalidating failed, the opened image will run in the no object map state, but at the same time it can continue writing. Actually, object map on the server is still valid, which will cause the mismatch between the object map and the actual image object and may cause image damage.

Now, that's a potential issue but the only solution would be to return IO errors in that case (or keep retrying to invalidate until successful if you hit a `ETIMEDOUT` error)

Songweibin commented:
> We shouldn't return an error to the app on IO -- and I don't see that in your appended log section:
>> 2020-03-19 11:13:36.327647 7f6017fff700 10 librbd::exclusive_lock::PostAcquireRequest: 0x7f6010003c10 handle_open_object_map: r=0

```
2020-03-19 11:13:26.294961 7f6017fff700 10 librbd::object_map::RefreshRequest: 0x7f6010004570 handle_load: r=-110
// failed to open object map
2020-03-19 11:13:36.327620 7f6017fff700 -1 librbd::object_map::InvalidateRequest: 0x7f60080689b0 should_complete: r=-110
// failed to invalidate object map, error was not returned to `handle_open_object_map`
```

@dillaman  Modified and repushed, please help to have a look. Thank you.

Now, if timed out to open object map when APP try to acquire exclusive lock (via `librbd::lock_acquire()`), donot set object map invalid and return -ETIMEDOUT to APP.

Songweibin commented:
@dillaman Hi, please help to have a look, thank you.

dillaman commented on src/librbd/object_map/RefreshRequest.cc:
```diff
@@ -133,6 +134,15 @@ Context *RefreshRequest<I>::handle_load(int *ret_val) {
+    m_image_ctx.owner_lock.lock_shared();
+    if (*ret_val == -ETIMEDOUT &&
+        !m_image_ctx.get_exclusive_lock_policy()->may_auto_request_lock() &&
+        cct->_conf.get_val<bool>("rbd_auto_exclusive_lock_until_manual_request")) {
```
I'd rather see a new developer-level configuration option to enable this behavior (since it's really corner-case). I would also like to see the invalidate issue fixed as well so that the `-ETIMEDOUT` error propagates back in the normal case.

Songweibin commented:
@dillaman  Updated, please have a review, thank you.

dillaman commented on src/common/options.cc:
```diff
@@ -7225,6 +7225,10 @@ static std::vector<Option> get_rbd_options() {
     .set_default(true)
     .set_description("validate new image names for RBD compatibility"),
 
+    Option("rbd_validate_object_map", Option::TYPE_BOOL, Option::LEVEL_DEV)
```
Nit: perhaps this should be `rbd_invalidate_object_map_on_load_timeout`. It's not "validating" the object map and it's only for that one specific case.

dillaman commented on src/test/librbd/object_map/test_mock_InvalidateRequest.cc:
```diff
@@ -44,7 +44,7 @@ TEST_F(TestMockObjectMapInvalidateRequest, UpdatesInMemoryFlag) {
     request->send();
   }
-  ASSERT_EQ(0, cond_ctx.wait());
+  ASSERT_EQ(-EROFS, cond_ctx.wait());
```
Hmm -- where is `-EROFS` coming from? I would want this test to expect no error. Same comment applies below.

Songweibin replied:
@dillaman  It comes from [L56](https://github.com/ceph/ceph/blob/master/src/librbd/object_map/InvalidateRequest.cc#L56) (since error have propagated back). 
I passed a param of [true](https://github.com/ceph/ceph/pull/34059/commits/f1386ebc0fc349405f7579584054df1255788dee#diff-def0606534f993f6de4ad41c2ece775cR36) so that it doesn't return an error, but i'm not sure if you like it, please take a look.  -- All updated. Thank you.

dillaman commented on src/test/librbd/object_map/test_mock_InvalidateRequest.cc:
```diff
@@ -144,7 +144,7 @@ TEST_F(TestMockObjectMapInvalidateRequest, IgnoresOnDiskUpdateFailure) {
     request->send();
   }
-  ASSERT_EQ(0, cond_ctx.wait());
+  ASSERT_EQ(-EINVAL, cond_ctx.wait());
```
Nit: the name of the test should be updated.

dillaman commented on src/test/librbd/object_map/test_mock_SnapshotCreateRequest.cc:
```diff
@@ -130,7 +130,7 @@ TEST_F(TestMockObjectMapSnapshotCreateRequest, ReadMapError) {
     request->send();
   }
-  ASSERT_EQ(0, cond_ctx.wait());
+  ASSERT_EQ(-ENOENT, cond_ctx.wait());
```
Nit: you need to tweak the test to ensure the object map can be invalidated w/o an `-ENOENT`. Same comment applies below.

dillaman replied:
Looks like this still needs to be addressed.

Songweibin replied:
@dillaman  Do you mean I should add tests w/o an `-ENOENT` in `expect_read_map`, and with `-ENOENT` in `expect_write_map` & `expect_add_snapshot`. Or would you help me how to do it? Thanks.

dillaman replied:
I mean that these tests shouldn't be receiving `-ENOENT` errors, so something needs to get fixed.

Songweibin replied:
@dillaman  Fix by never propagate an `-ENOENT` back to the caller.

A doubt: it seems that validating object map didn't take effect after `send_read_map/send_write_map/send_add_snapshot` failed, because it will return `-ENOENT` at [image_ctx.get_flags](https://github.com/ceph/ceph/blob/master/src/librbd/object_map/InvalidateRequest.cc#L30).

dillaman replied:
In the case where the snapshot object map was never created, we technically don't need to invalidate it since any attempt to load it will fail and mark it invalid at that point.

Songweibin replied:
Yeah, as you said, in the process of `object_map::SnapshotCreateRequest`, if read map fails, we technically don't need to invalidate it. So I think we expect an error (`ENOENT`) to be returned here is reasonable (as the tests name are ***Error and it always returns an `ENOENT` actually).
To be honest, I have no idea how to fix it, maybe passing `CEPH_NOSNAP` instead of `snap_id` at [SnapshotCreateRequest](https://github.com/ceph/ceph/pull/34059/commits/4d9aad06fedd57d84fa7c07b61214fba64230a19#diff-7561d125517840b2c7f07c4f7b817f4bR128) ? but I don't think it's reasonable.

dillaman commented on src/librbd/object_map/InvalidateRequest.h:
```diff
@@ -31,9 +31,12 @@ class InvalidateRequest : public AsyncRequest<ImageCtxT> {
-    // never propagate an error back to the caller
-    return 0;
+  int filter_return_code(int r) const override {
+    if (r == -ENOENT) {
```
I wasn't saying that we shouldn't propagate back the `-ENOENT` error, I was saying we should fix the cause of what was issuing the `-ENOENT` error.

Songweibin commented on src/librbd/object_map/Request.cc:
```diff
@@ -22,7 +22,10 @@ bool Request::should_complete(int r) {
   {
   case STATE_REQUEST:
-    if (r < 0) {
+    if (r == -ETIMEDOUT &&
```
@dillaman On the weekend, the update object map timeout occurred (there are slow request) when we were writing in 1M seq. In this corner-case, librbd does nothing but return IO errors to APPs, let it be handled by the APPs. Please help to take a look. Thanks.
```
2020-03-28 18:21:54.551307 7f5be97fa700 20 librbd::object_map::UpdateRequest: 0x7f5bdc047490 update_object_map: ictx=0x2b5f7e0, oid=rbd_object_map.14d156b8b4567, [24915,24916) = ->1
......
2020-03-28 18:22:25.551554 7f5be9ffb700 20 librbd::object_map::UpdateRequest: 0x7f5bdc047490 handle_update_object_map: r=-110
2020-03-28 18:22:25.551563 7f5be9ffb700 20 librbd::object_map::UpdateRequest: 0x7f5bdc047490 update_in_memory_object_map:
2020-03-28 18:22:25.551569 7f5be9ffb700 20 librbd::object_map::Request: 0x7f5bdc047490 should_complete: r=-110
2020-03-28 18:22:25.551575 7f5be9ffb700 -1 librbd::object_map::Request: failed to update object map: (110) Connection timed out
2020-03-28 18:22:25.551637 7f5be9ffb700 -1 librbd::object_map::InvalidateRequest: 0x7f5be00234c0 invalidating object map in-memory
2020-03-28 18:22:25.551643 7f5be9ffb700 -1 librbd::object_map::InvalidateRequest: 0x7f5be00234c0 invalidating object map on-disk
2020-03-28 18:22:25.569080 7f5be9ffb700 -1 librbd::object_map::InvalidateRequest: 0x7f5be00234c0 should_complete: r=0
2020-03-28 18:22:25.569111 7f5be9ffb700 20 librbd::object_map::Request: 0x7f5bdc047490 should_complete: r=0
2020-03-28 18:22:25.569113 7f5be9ffb700 20 librbd::object_map::Request: INVALIDATE
2020-03-28 18:22:25.569117 7f5be9ffb700 20 librbd::ObjectMap: 0x7f5bdc003820 handle_detained_aio_update: cell=0x7f5bdc172ff0, r=0
```

dillaman replied:
You should return `true` here and not invoke `finish_request` below. I also wonder if the "cure" is worse than the original problem since now your iSCSI clients will receive IO errors.

Songweibin replied:
Yup, I think iSCSI will handles this IO errors case. Updated, thank you.

dillaman replied:
Yeah, by bubbling errors back to the clients and having them report IO failures.

Songweibin replied:
Yes, in this case, we set `rados_osd_op_timeout = 31` (`osd_heartbeat_grace` + `osd_heartbeat_interval` + 5).

Songweibin commented:
Hi, @dillaman , is there anything else that needs to be modified? Thank you very much :-)

Songweibin commented:
@dillaman  Ping

dillaman commented:
I suspect this PR is responsible for the error here [1]:

```
2020-04-14T13:07:15.442 INFO:tasks.workunit.client.0.smithi069.stderr:======================================================================
2020-04-14T13:07:15.442 INFO:tasks.workunit.client.0.smithi069.stderr:ERROR: test_rbd.TestImage.test_set_snap_recreated
2020-04-14T13:07:15.442 INFO:tasks.workunit.client.0.smithi069.stderr:----------------------------------------------------------------------
2020-04-14T13:07:15.442 INFO:tasks.workunit.client.0.smithi069.stderr:Traceback (most recent call last):
2020-04-14T13:07:15.443 INFO:tasks.workunit.client.0.smithi069.stderr:  File "/usr/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
2020-04-14T13:07:15.443 INFO:tasks.workunit.client.0.smithi069.stderr:    self.test(*self.arg)
2020-04-14T13:07:15.443 INFO:tasks.workunit.client.0.smithi069.stderr:  File "/home/ubuntu/cephtest/clone.client.0/src/test/pybind/test_rbd.py", line 1089, in test_set_snap_recreated
2020-04-14T13:07:15.443 INFO:tasks.workunit.client.0.smithi069.stderr:    self.image.create_snap('snap1')
2020-04-14T13:07:15.444 INFO:tasks.workunit.client.0.smithi069.stderr:  File "rbd.pyx", line 3298, in rbd.requires_not_closed.wrapper
2020-04-14T13:07:15.444 INFO:tasks.workunit.client.0.smithi069.stderr:  File "rbd.pyx", line 3944, in rbd.Image.create_snap
2020-04-14T13:07:15.444 INFO:tasks.workunit.client.0.smithi069.stderr:rbd.ImageNotFound: [errno 2] RBD image not found (error creating snapshot b'snap1' from b'image160')
```

[1] http://qa-proxy.ceph.com/teuthology/jdillaman-2020-04-14_08:24:43-rbd-wip-jd-testing-2-distro-basic-smithi/4953429/teuthology.log

Songweibin commented:
Oh, sorry for this mistake, it is reproduced by allowing:
```
>>> image.create_snap('snap1')
>>> image.set_snap('snap1')
>>> image.remove_snap('snap1')
>>> image.create_snap('snap1')
```

debug log:
```
2020-04-16T19:24:58.739+0800 7f6722ffd700  5 librbd::object_map::SnapshotCreateRequest: 0x7f671000cf20 should_complete: state=WRITE_MAP, r=0
2020-04-16T19:24:58.739+0800 7f6722ffd700  5 librbd::object_map::SnapshotCreateRequest: 0x7f671000cf20 send_add_snapshot: oid=rbd_object_map.1013339262b7
2020-04-16T19:24:58.748+0800 7f6722ffd700  5 librbd::object_map::SnapshotCreateRequest: 0x7f671000cf20 should_complete: state=ADD_SNAPSHOT, r=-16
2020-04-16T19:24:58.748+0800 7f6722ffd700 20 librbd::object_map::Request: 0x7f671000cf20 should_complete: r=-16
2020-04-16T19:24:58.748+0800 7f6722ffd700 -1 librbd::object_map::Request: failed to update object map: (16) Device or resource busy
2020-04-16T19:24:58.748+0800 7f67227fc700 -1 librbd::object_map::InvalidateRequest: 0x7f6710009d30 should_complete: r=-2
// here, we have propagated back the error to the caller
2020-04-16T19:24:58.748+0800 7f67227fc700  5 librbd::object_map::SnapshotCreateRequest: 0x7f671000cf20 should_complete: state=ADD_SNAPSHOT, r=-2
2020-04-16T19:24:58.748+0800 7f67227fc700 20 librbd::object_map::Request: 0x7f671000cf20 should_complete: r=-2
2020-04-16T19:24:58.748+0800 7f67227fc700 20 librbd::object_map::Request: INVALIDATE
2020-04-16T19:24:58.748+0800 7f67227fc700 -1 librbd::object_map::Request: failed to invalidate object map: (2) No such file or directory
2020-04-16T19:24:58.748+0800 7f67227fc700  5 librbd::SnapshotCreateRequest: 0xc99320 handle_create_object_map: r=-2
2020-04-16T19:24:58.748+0800 7f67227fc700 -1 librbd::SnapshotCreateRequest: 0xc99320 handle_create_object_map: failed to snapshot object map: (2) No such file or directory
```

Fixed the test and updated. Thank you.

dillaman commented on src/test/pybind/test_rbd.py:
```diff
@@ -1082,7 +1082,7 @@ def test_set_snap_recreated(self):
         self.image.set_snap('snap1')
         self.image.remove_snap('snap1')
-        self.image.create_snap('snap1')
+        assert_raises(ImageNotFound, self.image.create_snap, 'snap1')
```
We can't change the existing behaviour of the API. We should be able to create a snapshot here.

Songweibin replied:
@dillaman  Yeah, I was thinking we could not propagate back the -ENOENT error in [InvalidateRequest](https://github.com/ceph/ceph/pull/34059/files#diff-1e91307bc8cf81a68ca4c24258347e5dL34) (but it seems you don't like it https://github.com/ceph/ceph/pull/34059#discussion_r399378444), so ...
I still prefer to not propagate back the -ENOENT error in InvalidateRequest, what do you think? thank you.

dillaman replied:
I thought that was covered by my comment here [1]. What's the source of this `-ENOENT`? 


[1] https://github.com/ceph/ceph/pull/34059#discussion_r399380790

Songweibin replied:
@dillaman  Removing an in-used (via `image.set_snap('snap1')`) snapshot will invalidated the image's object-map, and `image_ctx.get_flags(m_snap_id, &snap_flags)` [1] will return `-ENOENT` [2] when recreating a snapshot.

I have another way to resolve it as following (if you don't like this repushd fix [3], any help or suggestions would be appreciated.) Thank you inadvance.
```
--- a/src/librbd/object_map/Request.h
+++ b/src/librbd/object_map/Request.h
@@ -30,7 +30,8 @@ protected:

   bool should_complete(int r) override;
   int filter_return_code(int r) const override {
-    if (m_state == STATE_REQUEST) {
+    if (m_state == STATE_REQUEST ||
+        ((m_state == STATE_INVALIDATE) && r == -ENOENT)) {
       // never propagate an error back to the caller
       return 0;
     }
```

[1] [InvalidateRequest.cc#L30](https://github.com/ceph/ceph/blob/master/src/librbd/object_map/InvalidateRequest.cc#L30)
[2] [In the case where the snapshot object map was never created](https://github.com/ceph/ceph/pull/34059#discussion_r399380790)
[3] https://github.com/ceph/ceph/pull/34059/commits/374be22d267aaa57a3d081585c0fccef497651a5#diff-616fb4f060117fd81715aa3132a52454R58

Songweibin commented:
@dillaman  Ping.

dillaman commented:
It's on my radar -- I'm just backlogged at the moment.

Songweibin commented:
@dillaman Please help to have a look when you're convenient. Thank you :-)

dillaman commented:
jenkins test dashboard backend
